### PR TITLE
Open Hoot-UI data in josm

### PR DESCRIPTION
**Pull request by mzagorskirs**
Related Jira tickets:
- [VGI-3361](https://maxar.atlassian.net/browse/VGI-3361)
- [VGI-3401](https://maxar.atlassian.net/browse/VGI-3401)

Related PRs:
- [JOSM](https://github.com/DigitalGlobe/me-josm/pull/1494) (See this PR for discussion on updates)
- ^^^superseded by this PR: https://github.com/DigitalGlobe/me-josm/pull/1498

Demos:

https://user-images.githubusercontent.com/43011672/185269317-f863eab0-ce0b-4686-9b56-9590bbbc8648.mp4
https://user-images.githubusercontent.com/43011672/185269322-81c8f934-578a-4114-812f-5c2c2b877b52.mp4

### CODE REVIEW

**Comment by mzagorskirs** [comment]
Note: I will be updating the icon used in the modal.

**Comment by mzagorskirs** [comment]
Note: since this mechanism temporarily downloads the data first then deletes after alerting JOSM that there is a file to load, I'm not satisfied that this is the best approach.  I plan to look into utilizing JOSM's `import` API endpoint.

**Comment by brianhatchl** [comment]
I don't think there needs to be an export dialog at all for `Open in JOSM`.  Assume `include hoot tags` is true.

When the file download gets replaced by the remote control /import call, the url needs to looks something like
>http://127.0.0.1:8111/import?headers=Cookie,SESSION=eefcc756-4041-445f-a605-4ad78371bda9&url=https://localhost:8080/hoot-services/job/export/ex_65d5dfb9d78c4e7c9f141dd4a6efd466?outputname=Starbucks.osm.zip

I put together a test example in another branch by adding an action button to the Export jobs in the Jobs panel here https://github.com/ngageoint/hootenanny-ui/blob/85008687fd56f97499035ddc10da152889e4439f/modules/Hoot/managers/api.js#L982-L1000

File: modules/Hoot/managers/api.js [hunk 1]
```diff
@@ -1042,6 +1047,190 @@ export default class API {
...
+                    input = isDatasets ? d.map(n => n.name).join(',') : d.data.name,
+                    initialName = this.getOutputName(type, input),
+                    finalName = initialName.replace(/\s/g, '');
+
+                let data = {
+                    input: id,
+                    inputtype: this.getInputType(type),
+                    includehoottags: true,
+                    outputname: finalName,
+                    outputtype: 'OSM',
```
Comment (brianhatchl) [inline modules/Hoot/managers/api.js]: make this lowercase `osm`

File: modules/Hoot/managers/api.js [hunk 2]
```diff
@@ -1042,6 +1047,190 @@ export default class API {
...
+                this.openInJosm(data)
+                    .then(resp => {
+                        this.jobId = resp.data.jobid;
+
+                        return this.statusInterval(this.jobId);
+                    })
+                    .then(async resp => {
+
+                        if (resp.data && !this.isCancelled) {
+                            await this.openDatasetInJosm(this.jobId, data.outputname + '.OSM', data.outputname);
```
Comment (brianhatchl) [inline modules/Hoot/managers/api.js]: make this lowercase `.osm`

File: modules/Hoot/config/domMetadata.js [hunk 3]
```diff
@@ -4,6 +4,12 @@ import _cloneDeep from 'lodash-es/cloneDeep';
 import _isEmpty from 'lodash-es/isEmpty';
 import _isEqual from 'lodash-es/isEqual';
 
+/*
+* @apiNote Changelog: <br>
+*      Milla Zagorski 8-10-2022: Added code to allow for opening layer(s) in JOSM. <br>
+*
+*/
+
```
Comment (brianhatchl) [inline modules/Hoot/config/domMetadata.js]: this can be reverted if no other changes to file

File: modules/Hoot/managers/api.js [hunk 4]
```diff
@@ -1042,6 +1047,190 @@ export default class API {
...
+            return Promise.reject(new Error(' invalid request payload'));
+        }
+
+        const params = {
+            path: '/job/export/execute',
+            method: 'POST',
+            data: data
+        };
+        return this.request(params);
+    }
```
Comment (brianhatchl) [inline modules/Hoot/managers/api.js]: this is identical to `exportDataset` so should just use that below

File: modules/Hoot/managers/api.js [hunk 5]
```diff
@@ -1042,6 +1047,190 @@ export default class API {
...
+            default: break;
+        }
+        return type;
+    }
+
+    getTranslationPath(translations) {
+        let selectedTranslation = 'OSM';
+        const translation = translations.find(t => t.name === selectedTranslation);
+        return !translation.hasOwnProperty('path') ? translation.exportPath : translation.path;
+    }
```
Comment (brianhatchl) [inline modules/Hoot/managers/api.js]: these aren't really api methods, so would be better scoped as functions under `openDataInJosm`

File: modules/Hoot/ui/modals/exportData.js [hunk 6]
```diff
@@ -240,7 +240,7 @@ export default class ExportData {
             .then( async resp => {
                 if (resp.data && resp.data.status !== 'cancelled') {
                     await Hoot.api.saveDataset( this.jobId, data.outputname + '.' + data.outputtype + '.zip' );
-                    // or open in josm remote control
+                    // or open in JOSM remote control
                     //Hoot.api.openDatasetInJosm( this.jobId, data.outputname + '.' + data.outputtype );
```
Comment (brianhatchl) [inline modules/Hoot/ui/modals/exportData.js]: We should delete these two lines, they were just a placeholder when I was testing.

File: modules/Hoot/ui/managePanel/jobs.js [hunk 7]
```diff
@@ -960,10 +960,10 @@ export default class Jobs extends Tab {
                         }
                     });
                     actions.push({
-                        title: 'open in josm',
+                        title: 'open in JOSM',
```
Comment (brianhatchl) [inline modules/Hoot/ui/managePanel/jobs.js]: the tooltip convention I have going in the jobs panel is for them to be all lowercase, so this wasn't a typo

File: modules/Hoot/managers/api.js [hunk 8]
```diff
@@ -1042,6 +1047,190 @@ export default class API {
...
+                let data = {
+                    input: id,
+                    inputtype: this.getInputType(type),
+                    includehoottags: true,
+                    outputname: finalName,
+                    outputtype: 'OSM',
+                    translation: this.getTranslationPath(translations)
+                };
+
+                this.openInJosm(data)
```
Comment (brianhatchl) [inline modules/Hoot/managers/api.js]: could just use `exportData`

File: modules/Hoot/managers/api.js [hunk 9]
```diff
@@ -979,7 +982,7 @@ export default class API {
             });
     }
 
-    openDatasetInJosm( id, name ) {
+    openDatasetInJosm( id, name, outputName ) {
```
Comment (brianhatchl) [inline modules/Hoot/managers/api.js]: this signature would feel better to me as `(id, name, ext)` and then the ext could be appended on line 996 below.

File: modules/Hoot/managers/api.js [hunk 10]
```diff
@@ -988,6 +991,8 @@ export default class API {
             .then( resp => {
                 let rc = window.open('http://127.0.0.1:8111/import?'
                     + `headers=Cookie,SESSION=${resp.data}`
+                    + '&new_layer=true'
+                    + `&layer_name=${outputName}`
                     + `&url=${this.detect.origin}${this.baseUrl}/job/export/${id}?outputname=${name}.zip`
```
Comment (brianhatchl) [inline modules/Hoot/managers/api.js]: outputname=${name}.${ext}.zip

**Review by brianhatchl — CHANGES_REQUESTED** [review]

**Review by brianhatchl — APPROVED** [review]
LGTM :poultry_leg: :waffle:

**Pull request merged**